### PR TITLE
[kobo]: add reboot and poweroff menu actions

### LINKED ISSUE #2868: "Power off" function
* KOReader version: v2015.11-1023-g773e24f
* Device: Kobo Glo

#### Issue
From all the observations [of this post](https://www.mobileread.com/forums/showthread.php?p=3516659#post3516659), this is the only one that I can't find a workaround for.

Now that I can not use my "Turn off" button, I have to press the last menu icon>>>then the "Exit" entry>>>wait for koreader to exit>>>and press the "Shut Down" button of KSM to close the reader.
The reason that I don't use Koreader's shut down method is, as I wrote there:
"First is the faster wear of the power button.
 Second is that its really difficult because of the sliding nature of this button at the old Glo model. Its almost impossible to use it with gloves, and even without them, most of the time I miss it the first time.
 Another problem is that when the reader is closed or is sleeping, you see the same cover, with nothing to indicate the different states. You have to guess if you slide it for the right amount of time for it to go to the state that you intended."

So the proposition is to add a "Turn off" entry after the "Exit" one or even better have a way to add a top level button with that functionality.
The reason that I have to shut the reader down and not put it to sleep, is that it still drains the battery in suspend mode.

### PR DESCRIPTION
**Pull request by pazos**
**Don't merge it yet, I need to test it in my devices which aren't at home right now.**

Add menu entries for reboot and poweroff for kobos. I added those in common_info_menu_table.lua, but I'm not sure about the "info" thing :p

This should fix #2868 and #2588

### CODE REVIEW

File: frontend/ui/elements/common_info_menu_table.lua [hunk 1]
```diff
@@ -1,6 +1,8 @@
 local Device = require("device")
 local InfoMessage = require("ui/widget/infomessage")
 local UIManager = require("ui/uimanager")
+local Event = require("ui/event")
```
Comment (Frenzie) [inline frontend/ui/elements/common_info_menu_table.lua]: Event should go a little higher, see https://github.com/koreader/koreader-base/wiki/Coding-style#order-of-requires

**Comment by Frenzie** [comment]
I was wondering if it should be a submenu. These actions are a lot less harmless than regular exit, as in that case it still only takes seconds to restart the program.

Alternatively they could ask for confirmation.

Anyway, I'm not saying that's how it should be, but i do think it's something we should consciously do or not do.

**Comment by AlanSP1** [comment]
Before with xor_'s version (http://www.mobileread.com/forums/showpost.php?p=3148246&postcount=2294) shutting off was directly accessible in top menu row as last icon. It was enough to touch it. I never had problems with it, i.e. I never powered off device unintentionally.

If it is now in hamburger menu where exit is, I don't think it needs to be confirmed before.

**Comment by Frenzie** [comment]
In any case, I'm inclined to think the sleep/power off/restart should be grouped separately from exit.

Something like this:
```
------------------------
Suspend
Shut down
Restart
------------------------
Exit
```

**Comment by AlanSP1** [comment]
This looks just fine. :+1:

**Comment by KenMaltby** [comment]
Looks consistent with the way the menu separators should be used.  This, menu approach, should save some wear and tear on the Kobo power slide buttons.  It might even be nice if it could wake from a light button push, saving the power button altogether.

File: frontend/ui/elements/common_info_menu_table.lua [hunk 2]
```diff
@@ -49,5 +51,32 @@ common_info.report_bug = {
         })
     end
 }
+if Device:isKobo() then
+    common_info.reboot = {
+        text = _("Reboot"),
+    	  callback = function()
+            Screen:setRotationMode(0)
+            require("ui/screensaver"):show("reboot", _("Rebooting..."))
```
Comment (houqp) [inline frontend/ui/elements/common_info_menu_table.lua]: For translation, `…` is preferred over `...`.

File: frontend/ui/elements/common_info_menu_table.lua [hunk 3]
```diff
@@ -49,5 +51,32 @@ common_info.report_bug = {
...
+    common_info.poweroff = {
+        text = _("Power off"),
+    	  callback = function()
+            Screen:setRotationMode(0)
+            require("ui/screensaver"):show("poweroff", _("Powered off"))
+            Screen:refreshFull()
+            UIManager:nextTick(function()
+                UIManager:broadcastEvent(Event:new("Close"))
+                Device:powerOff()
+            end)
```
Comment (houqp) [inline frontend/ui/elements/common_info_menu_table.lua]: Can we reuse the function `poweroff_action` from UIManager instead? This, it's easier for us to push bug fix or update to this piece of code.

File: frontend/ui/elements/common_info_menu_table.lua [hunk 4]
```diff
@@ -49,5 +51,32 @@ common_info.report_bug = {
         })
     end
 }
+if Device:isKobo() then
+    common_info.reboot = {
+        text = _("Reboot"),
+    	  callback = function()
```
Comment (houqp) [inline frontend/ui/elements/common_info_menu_table.lua]: You probably want to set the `_entered_poweroff_stage` flag too here like what's done in `poweroff_action`. Bonus point, create a `reboot_action` function in UIManager and use it here ;)

File: frontend/device/kobo/device.lua [hunk 5]
```diff
@@ -517,6 +517,10 @@ function Kobo:powerOff()
     os.execute("poweroff")
 end
 
+function Kobo:reboot()
```
Comment (houqp) [inline frontend/device/kobo/device.lua]: I suggest we add a abstract `reboot` method in `generic/device.lua` too.

**Comment by houqp** [comment]
+1 on @Frenzie 's suggestion, but I prefer we name it `reboot` instead of `restart` to not confuse with @Hzj-jie 's restart patch.

File: frontend/ui/elements/common_info_menu_table.lua (hunk 4 above)
Comment (Frenzie) [inline frontend/ui/elements/common_info_menu_table.lua]: Also there's something odd about the indentation here. 4 spaces followed by a tab and then a space? Just stick with 4 spaces per indentation. :-)

File: frontend/ui/elements/common_info_menu_table.lua [hunk 6]
```diff
@@ -49,5 +51,32 @@ common_info.report_bug = {
...
+            UIManager:nextTick(function()
+                UIManager:broadcastEvent(Event:new("Close"))
+                Device:reboot()
+            end)
+        end
+    }
+
+    common_info.poweroff = {
+        text = _("Power off"),
+    	  callback = function()
```
Comment (Frenzie) [inline frontend/ui/elements/common_info_menu_table.lua]: indentation here as well

**Comment by KenMaltby** [comment]
I see "Restart" as restarting the current program, KOReader in this case, and "Reboot" as rebooting the device/OS.  Both can be valid steps to recover from different problems.

**Comment by AlanSP1** [comment]
Yes, as @KenMaltby commented, I too understand restart as software restart, and reboot as hardware restart.

**Comment by Hzj-jie** [comment]
Should we make the wording more explicit? i.e. "Reboot device" v.s. "Restart KOReader" and "Shutdown device" v.s. "Exit KOReader". "Suspend" does not really make too much confusing IMO: I cannot imagine the meaning of "suspend the software", debug trap?

**Comment by Frenzie** [comment]
I think @Hzj-jie makes a good point. Btw, it doesn't need to be called suspend. "Sleep" might make more sense given how we use that word elsewhere in the user-facing part of the device.

**Comment by KenMaltby** [comment]
Hmm... as long as the resulting action matches, i.e. "Reboot device" should do that, not "Restart KOReader" they are not the same.  "Exit KOReader" is not the same as "Shutdown device" or "Power Off".  Suspend and sleep are equivalent, as used here.

File: frontend/ui/elements/common_info_menu_table.lua (hunk 3 above)
Comment (pazos) [inline frontend/ui/elements/common_info_menu_table.lua]: yes, and call it with ```UIManager:scheduleIn(2,` self.poweroff_action)``` ????

File: frontend/ui/elements/common_info_menu_table.lua (hunk 2 above)
Comment (pazos) [inline frontend/ui/elements/common_info_menu_table.lua]: ok, thanks for that info!

File: frontend/device/kobo/device.lua (hunk 5 above)
Comment (pazos) [inline frontend/device/kobo/device.lua]: for logging purpose on emulator?

File: frontend/ui/elements/common_info_menu_table.lua (hunk 6 above)
Comment (pazos) [inline frontend/ui/elements/common_info_menu_table.lua]: meh, my atom setup seems broken for lua. Switching to zerobrane. Thanks for the info!

**Pull request closed without merging**